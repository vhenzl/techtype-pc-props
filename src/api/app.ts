import express from 'express';
import { createGetNodeSubtreeQueryHandler, GetNodeSubtreeQuerySchema } from '../application/use-cases/get-node-subtree.ts';
import { createConnection } from '../infrastructure/persistence/db.ts';
import { withQueryValidation } from '../infrastructure/processing.ts';
import { createGetSubtreeController } from './controllers.ts';
import * as middlewares from './middleware.ts';

// ----- services composition -----
const db = await createConnection();

const getSubtreeQueryHandler = withQueryValidation(
  GetNodeSubtreeQuerySchema,
  createGetNodeSubtreeQueryHandler({ db }),
);

const getSubtreeController = createGetSubtreeController(getSubtreeQueryHandler);

// ----- express app composition -----
const app = express();

app.use(middlewares.logger);
app.use(middlewares.requireJsonPost);
app.use(express.json());

app.use('/subtree/*path', getSubtreeController);

app.use(middlewares.errorLogger);
app.use(middlewares.errorHandler);

export default app;
