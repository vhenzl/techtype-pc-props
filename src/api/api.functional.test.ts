import express from 'express';
import request from 'supertest';
import { beforeAll, describe, expect, it } from 'vitest';
import { createApp } from './app.ts';

describe('API endpoint', () => {
  let app: express.Express;
  beforeAll(async () => {
    process.env['DATABASE_URL'] ??= import.meta.env.VITE_DATABASE_URL;
    app = await createApp();
  });

  describe('GET /subtree/*path', function () {
    it.for([
      ['/AlphaPC'],
      ['/AlphaPC/Processing/CPU'],
    ])('responds with expected json for path "%s"', async ([path]) => {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
      const res = await request(app).get(`/subtree${path}`).set('Accept', 'application/json');

      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
      expect(res.statusCode).toEqual(200);
      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
      expect(res.headers['content-type']).toMatch(/json/);
      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
      expect(res.body).toMatchSnapshot();
    });
  });
});
