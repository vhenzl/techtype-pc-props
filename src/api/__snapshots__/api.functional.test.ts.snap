// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`API endpoint > GET /subtree/*path > responds with expected json for path "/AlphaPC" 1`] = `
{
  "data": {
    "children": [
      {
        "children": [
          {
            "children": [],
            "id": "00000000-0000-7000-a000-100000000003",
            "name": "CPU",
            "parentId": "00000000-0000-7000-a000-100000000002",
            "properties": [
              {
                "id": "00000000-0000-7000-a000-200000000004",
                "name": "Cores",
                "value": 4,
              },
              {
                "id": "00000000-0000-7000-a000-200000000005",
                "name": "Power",
                "value": 2.41,
              },
            ],
          },
          {
            "children": [],
            "id": "00000000-0000-7000-a000-100000000004",
            "name": "Graphics",
            "parentId": "00000000-0000-7000-a000-100000000002",
            "properties": [
              {
                "id": "00000000-0000-7000-a000-200000000007",
                "name": "Ports",
                "value": 8,
              },
              {
                "id": "00000000-0000-7000-a000-200000000006",
                "name": "RAM",
                "value": 4000,
              },
            ],
          },
        ],
        "id": "00000000-0000-7000-a000-100000000002",
        "name": "Processing",
        "parentId": "00000000-0000-7000-a000-100000000001",
        "properties": [
          {
            "id": "00000000-0000-7000-a000-200000000003",
            "name": "RAM",
            "value": 32000,
          },
        ],
      },
      {
        "children": [
          {
            "children": [],
            "id": "00000000-0000-7000-a000-100000000007",
            "name": "HDD",
            "parentId": "00000000-0000-7000-a000-100000000005",
            "properties": [
              {
                "id": "00000000-0000-7000-a000-200000000010",
                "name": "Capacity",
                "value": 5120,
              },
              {
                "id": "00000000-0000-7000-a000-200000000011",
                "name": "WriteSpeed",
                "value": 1.724752,
              },
            ],
          },
          {
            "children": [],
            "id": "00000000-0000-7000-a000-100000000006",
            "name": "SSD",
            "parentId": "00000000-0000-7000-a000-100000000005",
            "properties": [
              {
                "id": "00000000-0000-7000-a000-200000000008",
                "name": "Capacity",
                "value": 1024,
              },
              {
                "id": "00000000-0000-7000-a000-200000000009",
                "name": "WriteSpeed",
                "value": 250,
              },
            ],
          },
        ],
        "id": "00000000-0000-7000-a000-100000000005",
        "name": "Storage",
        "parentId": "00000000-0000-7000-a000-100000000001",
        "properties": [],
      },
    ],
    "id": "00000000-0000-7000-a000-100000000001",
    "name": "AlphaPC",
    "parentId": null,
    "properties": [
      {
        "id": "00000000-0000-7000-a000-200000000001",
        "name": "Height",
        "value": 450,
      },
      {
        "id": "00000000-0000-7000-a000-200000000002",
        "name": "Width",
        "value": 180,
      },
    ],
  },
}
`;

exports[`API endpoint > GET /subtree/*path > responds with expected json for path "/AlphaPC/Processing/CPU" 1`] = `
{
  "data": {
    "children": [],
    "id": "00000000-0000-7000-a000-100000000003",
    "name": "CPU",
    "parentId": "00000000-0000-7000-a000-100000000002",
    "properties": [
      {
        "id": "00000000-0000-7000-a000-200000000004",
        "name": "Cores",
        "value": 4,
      },
      {
        "id": "00000000-0000-7000-a000-200000000005",
        "name": "Power",
        "value": 2.41,
      },
    ],
  },
}
`;
